text/0000-generic-integers.md

```diff
@@ -0,0 +1,437 @@
+- Feature Name: `generic_integers`
+- Start Date: 2024-08-07
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Adds the builtin types `u<N>` and `i<N>`, allowing integers with an arbitrary size in bits.
+
+# Motivation
+[motivation]: #motivation
+
+## Generalising code for integers
+
+Right now, there's a *lot* of boilerplate for implementing methods for integer primitives. The standard library itself is a great example; almost the entirety of `core::num` uses some gnarly macros to define all sorts of traits for all the
+integer types. One example is `Shl` and `Shr`, which are defined for not just every integer type, but every *combination* of integer types. We could easily do this with const generics instead:
+
+```rust
+impl<const N: usize, const M: usize> Shl<u<M>> for u<N> {
+    type Output = u<N>;
+    #[inline]
+    fn shl(self, rhs: u<M>) -> u<N> {
+        // implementation
+    }
+}
+```
+
+This will decrease compilation time across the entire Rust ecosystem, maybe not by a noticeable amount, but by some amount, due to the presence of macro-based trait implementations across the board.
+
+## Documentation decluttering
+
+Having generic impls would drastically reduce the noise in the "implementations" section of rustdoc. For example, the number of implementations for `Add` for integer types really drowns out the fact that it's also implemented for strings and `std::time` types, which is useful to know too.
+
+## Enum optimisations
+
+Using a smaller number of bits for a value also has the benefit of niche enum optimisations. For example, `u<7>` represents a single ASCII character, and `Option<u<7>>` can be stored in a single byte. Additionally, `Result<u<7>, E>` also takes one byte if `E` is zero-sized.
+
+This works because we know that values in the range `0b0_0000000..=0b0_1111111` are valid `u<7>`, but the values in the range `0b1_0000000..=0b1_1111111` are not and can represent the zero-sized variants `None` and `Err(E)`.
+
+## Bit masks
+
+Integers are very useful as a simple list of bits, and specifically for generic integers, this allows numbers of bits that aren't an existing integer type. There will probably always be a need for dedicated data structures like [`BitVec`], but at least for simple cases, being able to do this with your standard integer types is nice too.
```

Comment:
This is bad argument, too, because this is usually handled by `bitflags` crate. You don't explain here how your approach is better than `bitflags`. So, please, start with _truly_ convincing arguments, then move to others

Reply:
First, please be less dismissive of the author. Second, I don't understand why you're referencing `bitflags` here. That crate is appropriate when you have a fixed-at-design-time number of bit masks that you can sensibly give Rust identifiers to. In contrast, this section talks about the N-element bit vectors where N is may be a const parameter or it otherwise isn't sensible to give a distinct Rust identifier to each individual bit (e.g., because they correspond to `usize` indices in some other array or lanes in a SIMD vector). You can tell that this is what the RFC means because it *itself* references an existing crate that does what's being discussed here, `bitvec`, which solves a rather different problem from `bitflags`.

Reply:
Yeah, all of these different data structures have different meanings, and the main thing I was going for here is specifically on variable-size bit sequences. This proposal does not include any "bigint" types with arbitrary amounts of bits, and so, if you want such a structure, you're going to have to rely on the crate ecosystem for that.

`bitflags` and similar crates will still exist after this, as usual. Although it's worth noting that bitflags could be easily improved by these extra integer types, since they could easily allow niches for wrappers if the number of bits needed is less than a power of two.